### PR TITLE
Convention set builder refactoring

### PR DESCRIPTION
**Pull request by ajcvickers**
Issue #10672

Now we have:
* `IProviderConventionSetBuilder` -> `ProviderConventionSetBuilder` -> `[RelationalConventionSetBuilder (for relational providers)]` -> `<MyProviderConventionSetBuilder>`
  * Single registration, scoped service with all the conventions supplied/modified by the provider
* `IConventionSetCustomizer` -> `<MyConventionSetCustomizer>`
  * Multiple registration; allows other extensions to customize conventions
  * Currently used by the proxies package
* `IConventionSetBuilder` -> `RuntimeConventionSetBuilder`
  * The combined set of provider conventions with customizations
  * This is the service that should be resolved to get the current set of conventions being used

### CODE REVIEW

File: src/EFCore/Metadata/Conventions/Internal/IConventionSetCustomizer.cs [hunk 1]
```diff
@@ -13,17 +11,20 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal
...
+    ///         The service lifetime is <see cref="ServiceLifetime.Scoped"/> and multiple registrations
+    ///         are allowed. This means that each <see cref="DbContext"/> instance will use its own
+    ///         set of instances of this service.
+    ///         The implementations may depend on other services registered with any lifetime.
+    ///         The implementations do not need to be thread-safe.
     ///     </para>
     /// </summary>
-    public interface ICoreConventionSetBuilder
+    // Issue#11266 This interface is being used by provider code. Do not break.
+    public interface IConventionSetCustomizer
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/Internal/IConventionSetCustomizer.cs]: Might as well move this to `.Infrastructure`

File: src/EFCore/Metadata/Conventions/Internal/IProviderConventionSetBuilder.cs [hunk 2]
```diff
@@ -18,13 +18,13 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal
     ///         The implementations do not need to be thread-safe.
     ///     </para>
     /// </summary>
-    public class NullConventionSetBuilder : IConventionSetBuilder
+    // Issue#11266 This interface is being used by provider code. Do not break.
+    public interface IProviderConventionSetBuilder
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/Internal/IProviderConventionSetBuilder.cs]: Move this to `.Infrastructure` too

File: src/EFCore/Metadata/Conventions/Internal/IConventionSetCustomizer.cs (hunk 1 above)
Comment (ajcvickers) [reply src/EFCore/Metadata/Conventions/Internal/IConventionSetCustomizer.cs]: Will do on both.

File: src/EFCore/Metadata/Conventions/Internal/IProviderConventionSetBuilder.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Conventions/Internal/IProviderConventionSetBuilder.cs]: `ProviderConventionSetBuilder` and `RelationalConventionSetBuilder` as well

File: src/EFCore/Metadata/Conventions/Infrastructure/IConventionSetCustomizer.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    ///         The implementations do not need to be thread-safe.
+    ///     </para>
+    /// </summary>
+    public interface IConventionSetCustomizer
+    {
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        ConventionSet ModifyConventions(ConventionSet conventionSet);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/Infrastructure/IConventionSetCustomizer.cs]: Update doc comment

File: src/EFCore/Metadata/Conventions/Infrastructure/ProviderConventionSetBuilder.cs [hunk 4]
```diff
@@ -1,32 +1,44 @@
...
-    public class CoreConventionSetBuilder : ICoreConventionSetBuilder
+    public class ProviderConventionSetBuilder : IProviderConventionSetBuilder
     {
         /// <summary>
-        ///     Initializes a new instance of the <see cref="CoreConventionSetBuilder" /> class.
+        ///     Initializes a new instance of the <see cref="ProviderConventionSetBuilder" /> class.
         /// </summary>
         /// <param name="dependencies"> Parameter object containing dependencies for this service. </param>
-        public CoreConventionSetBuilder([NotNull] CoreConventionSetBuilderDependencies dependencies)
+        public ProviderConventionSetBuilder([NotNull] CoreConventionSetBuilderDependencies dependencies)
```
Comment (smitpatel) [inline src/EFCore/Metadata/Conventions/Infrastructure/ProviderConventionSetBuilder.cs]: Dependency object name should be changed.

Comment (ajcvickers) [reply src/EFCore/Metadata/Conventions/Infrastructure/ProviderConventionSetBuilder.cs]: Good catch!

**Pull request merged**